src/cmd/provision.rs

```diff
@@ -0,0 +1,194 @@
+        /// The name of the environment
+        #[arg(short = 'n', long)]
+        name: String,
+    },
+    /// Provision generic nodes for an environment
+    #[clap(name = "generic-nodes")]
+    GenericNodes {
+        /// The name of the environment
+        #[arg(short = 'n', long)]
+        name: String,
+    },
+    /// Provision peer cache nodes for an environment
+    #[clap(name = "peer-cache-nodes")]
+    PeerCacheNodes {
+        /// The name of the environment
+        #[arg(short = 'n', long)]
+        name: String,
+    },
+    /// Provision symmetric private nodes for an environment
+    #[clap(name = "symmetric-private-nodes")]
+    SymmetricPrivateNodes {
+        /// The name of the environment
+        #[arg(short = 'n', long)]
+        name: String,
+    },
+    /// Provision uploader nodes for an environment
+    #[clap(name = "uploaders")]
+    Uploaders {
+        /// The name of the environment
+        #[arg(short = 'n', long)]
+        name: String,
+    },
+}
+
+async fn init_provision(
+    name: &str,
+) -> Result<(
+    DeployOptions,
+    ProvisionOptions,
+    sn_testnet_deploy::ansible::provisioning::AnsibleProvisioner,
+    sn_testnet_deploy::ssh::SshClient,
+)> {
+    let deploy_options: DeployOptions = get_options_from_s3(name, OptionsType::Deploy).await?;
+    let provision_options: ProvisionOptions =
+        get_options_from_s3(name, OptionsType::Provision).await?;
+
+    let mut builder = TestnetDeployBuilder::default();
+    builder
+        .ansible_verbose_mode(false)
+        .deployment_type(deploy_options.environment_type.clone())
+        .environment_name(name)
+        .provider(CloudProvider::DigitalOcean);
+    let testnet_deployer = builder.build()?;
+
+    let inventory_service = DeploymentInventoryService::from(&testnet_deployer);
+    inventory_service
+        .generate_or_retrieve_inventory(name, true, Some(deploy_options.binary_option.clone()))
+        .await?;
+
+    let provisioner = testnet_deployer.ansible_provisioner;
+    let ssh_client = testnet_deployer.ssh_client;
+
+    Ok((deploy_options, provision_options, provisioner, ssh_client))
+}
+
+async fn handle_provision_nodes(name: String, node_type: NodeType) -> Result<()> {
+    let (deploy_options, mut provision_options, provisioner, ssh_client) =
+        init_provision(&name).await?;
+
+    let (genesis_multiaddr, genesis_ip) =
+        get_genesis_multiaddr(&provisioner.ansible_runner, &ssh_client).map_err(|err| {
+            println!("Failed to get genesis multiaddr {err:?}");
+            err
+        })?;
+    let genesis_network_contacts = get_bootstrap_cache_url(&genesis_ip);
+
+    let private_node_inventory = PrivateNodeProvisionInventory::new(
+        &provisioner,
+        deploy_options.full_cone_private_node_vm_count,
+        deploy_options.symmetric_private_node_vm_count,
+    )?;
+
+    match node_type {
+        NodeType::FullConePrivateNode | NodeType::SymmetricPrivateNode => match node_type {
```

Comment:
We could combine these into a single match statement, don't need for 2. Also it would be nice to print something to tell that we couldn't provision since `private_node_inventory.should_provision_full_cone_private_nodes()` was false, this happens if the `*private_node_vm_count` was set to 0 during the initial to_deploy.